libipt/src/pt_block_decoder.c

```diff
@@ -2515,7 +2521,14 @@ pt_blk_proceed_no_event_fill_cache(struct pt_block_decoder *decoder,
 	 * Cache updates are atomic so even if the two versions were not
 	 * identical, we wouldn't care because they are both correct.
 	 */
-	return pt_bcache_add(bcache, ioff, bce);
+	status = pt_bcache_add(bcache, ioff, bce);
+	if (status < 0)
+		return status;
+
+	if (valid_cache)
```

Comment:
We'd need some comment explaining that we can only proceed with the new cache entry if we have not filled the block.